sympy/simplify/combsimp.py

```diff
@@ -64,450 +47,82 @@ def combsimp(expr):
-        if level == 2:
-            sifted = sift(expr.args, gamma_factor)
-            gamma_ind = Mul(*sifted.pop(False, []))
-            d = Mul(*sifted.pop(True, []))
-            assert not sifted
+    expr = _gammasimp(expr, as_comb = True)
+    expr = _gamma_as_comb(expr)
+    return expr
 
-            nd, dd = d.as_numer_denom()
-            for ipass in range(2):
-                args = list(ordered(Mul.make_args(nd)))
-                for i, ni in enumerate(args):
-                    if ni.is_Add:
-                        ni, dd = Add(*[
-                            rule_gamma(gamma_rat(a/dd), level + 1) for a in ni.args]
-                            ).as_numer_denom()
-                        args[i] = ni
-                        if not dd.has(gamma):
-                            break
-                nd = Mul(*args)
-                if ipass ==  0 and not gamma_factor(nd):
-                    break
-                nd, dd = dd, nd  # now process in reversed order
-            expr = gamma_ind*nd/dd
-            if not (expr.is_Mul and (gamma_factor(dd) or gamma_factor(nd))):
-                return expr
-            level += 1
 
-        # iteration until constant
-        if level == 3:
-            while True:
-                was = expr
-                expr = rule_gamma(expr, 4)
-                if expr == was:
-                    return expr
+def _gamma_as_comb(expr):
+    """
+    Helper function for combsimp.
 
-        numer_gammas = []
-        denom_gammas = []
-        numer_others = []
-        denom_others = []
-        def explicate(p):
-            if p is S.One:
-                return None, []
-            b, e = p.as_base_exp()
-            if e.is_Integer:
-                if b.func is gamma:
-                    return True, [b.args[0]]*e
-                else:
-                    return False, [b]*e
-            else:
-                return False, [p]
+    Rewrites expression in terms of factorials and binomials
+    """
 
-        newargs = list(ordered(expr.args))
-        while newargs:
-            n, d = newargs.pop().as_numer_denom()
-            isg, l = explicate(n)
-            if isg:
-                numer_gammas.extend(l)
-            elif isg is False:
-                numer_others.extend(l)
-            isg, l = explicate(d)
-            if isg:
-                denom_gammas.extend(l)
-            elif isg is False:
-                denom_others.extend(l)
+    expr = expr.rewrite(factorial)
+
+    from .simplify import bottom_up
+
+    def f(rv):
+        n, d = rv.as_numer_denom()
+
+        n_args = []
+        n_fact_args = []
+        if isinstance(n, (Mul, Pow, factorial)):
+            for factor, exp in n.as_powers_dict().items():
+                try:
+                    n_args.extend([factor]*exp)
+                    if isinstance(factor, factorial):
+                        n_fact_args.extend([factor.args[0]]*exp)
+                except TypeError:
```

Comment:
Under which conditions does this occur?

Reply:
@jkusom If exp is non-integer(not int castable).